hy/scoping.py

```diff
@@ -0,0 +1,243 @@
+import ast
+import itertools
+
+from hy.lex import mangle
+from hy.models import Expression, List, Symbol
+
+import hy._compat
+
+
+class NodeWithName:
+    def __init__(self, node, index=None):
+        self.node = node
+        self.index = index
+
+    @property
+    def name(self):
+        if isinstance(self.node, (ast.Global, ast.Nonlocal)):
+            return self.node.names[self.index]
+        elif hy._compat.PY3_10:
+            if isinstance(self.node, (ast.MatchAs, ast.MatchStar)):
+                return self.node.name
+            elif isinstance(self.node, ast.MatchMapping):
+                return self.node.rest
+        return self.node.id
+
+    @name.setter
+    def name(self, new_name):
+        if isinstance(self.node, (ast.Global, ast.Nonlocal)):
+            self.node.names[self.index] = new_name
+        elif hy._compat.PY3_10:
+            if isinstance(self.node, (ast.MatchAs, ast.MatchStar)):
+                self.node.name = new_name
+            elif isinstance(self.node, ast.MatchMapping):
+                self.node.rest = new_name
+        self.node.id = new_name
+
+
+class Scope:
+    def __init__(self, stack, scope_id):
+        self.stack = stack
+        self.id = scope_id
+
+    def __enter__(self):
+        self.parent = self.stack.scope
```

Comment:
parent isn't defined in `__init__` and will raise an unexpected exception in certain edge cases. i.e.

```clojure
=> (let [a 1 b 2] (nonlocal a) (global b))
Traceback (most recent call last):
  File "stdin-bf89d46423db3acec3100628e39c19de80650ffb", line 1, in <module>
    (let [a 1 b 2] (nonlocal a) (global b))
hy.errors.HyMacroExpansionError:
  File "<stdin>", line 1
    (let [a 1 b 2] (nonlocal a) (global b))
                   ^----------^
expanding macro nonlocal
  AttributeError: 'Scope' object has no attribute 'parent'
```
`self.parent` should be initialized in `__init__` to `None`